### ISSUE TITLE
Add support for PyTorch Geometric-style `edge_index` graph input

### ISSUE THREAD

**Issue by antoneri** [mapequation/infomap#536]
Graph ML users often represent graphs using PyTorch Geometric-style tensors:

```python
edge_index.shape == (2, num_edges)
```

where:

- `edge_index[0]` contains source node IDs,
- `edge_index[1]` contains target node IDs.

Even when users are not using PyTorch Geometric directly, this representation has become a common interchange format for graph learning pipelines.

Supporting it would make Infomap easier to use for graph ML baselines, community-aware pooling, preprocessing, label generation, and evaluation.

## Goal

Implement a public Python helper that constructs an Infomap instance from an `edge_index` representation, with optional edge weights and node count.

## Proposed API

```python
from infomap import Infomap

im = Infomap.from_edge_index(
    edge_index,
    edge_weight=None,
    num_nodes=None,
    directed=True,
    node_ids=None,
    args=None,
)

im.run()
modules = im.get_modules()
```

Alternative acceptable API:

```python
from infomap.adapters import from_edge_index

im = from_edge_index(edge_index, edge_weight=edge_weight)
```

## Supported input types

`edge_index` should accept:

- NumPy array with shape `(2, m)`.
- PyTorch tensor with shape `(2, m)`, if PyTorch is installed.
- List or tuple convertible to a NumPy array with shape `(2, m)`.

`edge_weight` should accept:

- `None`.
- NumPy array of shape `(m,)`.
- PyTorch tensor of shape `(m,)`, if PyTorch is installed.
- Python list of numeric values of length `m`.

The implementation must not require PyTorch as a mandatory dependency.

If PyTorch tensors are provided, support them by converting via:

```python
tensor.detach().cpu().numpy()
```

### `edge_index`

Two-row array-like structure where row 0 contains source node IDs and row 1 contains target node IDs.

### `edge_weight: Optional[array-like] = None`

Optional edge weights.

If omitted, every edge should have weight `1.0`.

### `num_nodes: Optional[int] = None`

Total number of nodes.

If omitted, infer as:

```python
max(edge_index) + 1
```

for integer node IDs.

If the graph contains isolated nodes, users must pass `num_nodes`.

### `directed: bool = True`

Whether to treat edges as directed.

For `directed=False`, the implementation should avoid double-counting duplicate reverse edges where possible and document the behavior.

### `node_ids: Optional[Sequence[Hashable]] = None`

Optional external node IDs.

If provided, `node_ids[i]` maps internal integer node ID `i` to an external node ID.

### `args: Optional[str] = None`

Optional Infomap argument string passed to the underlying `Infomap` constructor.

## Validation

Raise `ValueError` if:

- `edge_index` does not have shape `(2, m)`.
- `edge_weight` is provided and does not have length `m`.
- `num_nodes` is smaller than the maximum node index plus one.
- node indices are negative.
- node indices are non-integer.
- `node_ids` is provided and has length different from `num_nodes`.
- weights contain NaN.
- weights contain negative values, unless Infomap explicitly supports them.

## Optional PyTorch Geometric helper

If PyTorch Geometric is installed, support direct construction from a `torch_geometric.data.Data` object:

```python
im = Infomap.from_pyg_data(
    data,
    edge_weight_attr="edge_weight",
    directed=True,
    args=None,
)
```

Expected behavior:

- read `data.edge_index`,
- use `data.edge_weight` if present and `edge_weight_attr="edge_weight"`,
- use `data.num_nodes` if present, so isolated nodes are preserved,
- do not require PyG as a dependency unless this method is used,
- if PyG is not installed and the user calls `from_pyg_data`, raise `ImportError` with a clear message.

This helper can be implemented in the same issue or as a follow-up. The core requirement is `from_edge_index`.

## Examples

Basic NumPy example:

```python
import numpy as np
from infomap import Infomap

edge_index = np.array([
    [0, 1, 1, 2],
    [1, 0, 2, 1],
])

im = Infomap.from_edge_index(edge_index, directed=False)
im.run()

modules = im.get_modules()
assert set(modules.keys()) == {0, 1, 2}
```

Weighted example:

```python
import numpy as np
from infomap import Infomap

edge_index = np.array([
    [0, 1, 1, 2],
    [1, 0, 2, 1],
])

edge_weight = np.array([1.0, 1.0, 3.5, 3.5])

im = Infomap.from_edge_index(
    edge_index,
    edge_weight=edge_weight,
    directed=False,
)
```

PyTorch tensor example:

```python
import torch
from infomap import Infomap

edge_index = torch.tensor([
    [0, 1, 2],
    [1, 2, 0],
])

im = Infomap.from_edge_index(edge_index, directed=True)
```

## Tests

Add tests for:

- NumPy `edge_index`.
- list-based `edge_index`.
- PyTorch tensor input, gated behind an optional dependency marker.
- weighted edges.
- unweighted edges.
- directed graph.
- undirected graph.
- isolated nodes with `num_nodes`.
- custom `node_ids`.
- invalid shape.
- invalid weights length.
- negative node ID.
- non-integer node ID.
- `num_nodes` too small.
- PyG `Data` input, if PyG is available in the test environment or behind an optional test marker.

## Documentation

Add documentation section:

```text
Python API: edge_index and PyTorch Geometric workflows
```

Include:

- `edge_index` example,
- weighted example,
- PyTorch tensor example,
- PyG `Data` example if implemented,
- note that PyTorch and PyG are optional dependencies,
- note on preserving isolated nodes via `num_nodes`.

## Acceptance criteria

- Users can construct Infomap from PyG-style `edge_index`.
- PyTorch is not required for standard use.
- PyTorch tensors are supported when PyTorch is installed.
- Optional PyG `Data` support works if implemented.
- Isolated nodes are preserved when `num_nodes` is provided.
- Tests cover validation and expected module assignment keys.
- Documentation includes runnable examples.
